api/args/linkedin/profile/profile.go

```diff
@@ -0,0 +1,141 @@
+package profile
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/masa-finance/tee-worker/api/args/base"
+	"github.com/masa-finance/tee-worker/api/types"
+	"github.com/masa-finance/tee-worker/api/types/linkedin/experiences"
+	"github.com/masa-finance/tee-worker/api/types/linkedin/functions"
+	"github.com/masa-finance/tee-worker/api/types/linkedin/industries"
+	"github.com/masa-finance/tee-worker/api/types/linkedin/profile"
+	"github.com/masa-finance/tee-worker/api/types/linkedin/seniorities"
+)
+
+var (
+	ErrScraperModeNotSupported = errors.New("scraper mode not supported")
+	ErrMaxItemsTooLarge        = errors.New("max items must be less than or equal to 100")
+	ErrExperienceNotSupported  = errors.New("years of experience not supported")
+	ErrSeniorityNotSupported   = errors.New("seniority level not supported")
+	ErrFunctionNotSupported    = errors.New("function not supported")
+	ErrIndustryNotSupported    = errors.New("industry not supported")
+	ErrUnmarshalling           = errors.New("failed to unmarshal LinkedIn profile arguments")
+)
+
+const (
+	DefaultMaxItems    = 10
+	DefaultScraperMode = profile.ScraperModeShort
+	MaxItems           = 1000 // 2500 on the actor, but we will run over 1MB memory limit on responses
+)
+
+// Verify interface implementation
+var _ base.JobArgument = (*Arguments)(nil)
+
+// Arguments defines args for LinkedIn profile operations
+type Arguments struct {
+	Type                  types.Capability    `json:"type"`
+	ScraperMode           profile.ScraperMode `json:"profileScraperMode"`
+	Query                 string              `json:"searchQuery"`
+	MaxItems              uint                `json:"maxItems"`
+	Locations             []string            `json:"locations,omitempty"`
+	CurrentCompanies      []string            `json:"currentCompanies,omitempty"`
+	PastCompanies         []string            `json:"pastCompanies,omitempty"`
+	CurrentJobTitles      []string            `json:"currentJobTitles,omitempty"`
+	PastJobTitles         []string            `json:"pastJobTitles,omitempty"`
+	Schools               []string            `json:"schools,omitempty"`
+	YearsOfExperience     []experiences.Id    `json:"yearsOfExperienceIds,omitempty"`
+	YearsAtCurrentCompany []experiences.Id    `json:"yearsAtCurrentCompanyIds,omitempty"`
+	SeniorityLevels       []seniorities.Id    `json:"seniorityLevelIds,omitempty"`
+	Functions             []functions.Id      `json:"functionIds,omitempty"`
+	Industries            []industries.Id     `json:"industryIds,omitempty"`
+	FirstNames            []string            `json:"firstNames,omitempty"`
+	LastNames             []string            `json:"lastNames,omitempty"`
+	RecentlyChangedJobs   bool                `json:"recentlyChangedJobs,omitempty"`
+	StartPage             uint                `json:"startPage,omitempty"`
+}
+
+func (t *Arguments) UnmarshalJSON(data []byte) error {
```

Comment:
...and this sort of thing is why we ask "Why doesn't Go have metaprogramming"? Good that we're using the same pattern everywhere!

I wonder... should we have a document (perhaps in the README or somewhere else) where we document the patterns that we use and, in cases like this, why we use them? It initially looks like a WTF until you either think about it or someone explains it.

Reply:
Yeah this is where interfaces come in but the repetition gets annoying for sure.  I will add a note in the readme to document some patterns